src/box/space.c

```diff
@@ -237,6 +238,56 @@ space_cleanup_constraints(struct space *space)
 	return 0;
 }
 
+/**
+ * Pin collation identifier with `id` in the cache, so that it can't be deleted.
+ */
+static void
+space_pin_collations_helper(struct space *space, uint32_t id,
+			    enum coll_id_holder_type holder_type)
+{
+	struct coll_id *coll_id = coll_by_id(id);
+	assert(coll_id != NULL);
+	struct coll_id_cache_holder *h = xmalloc(sizeof(*h));
+	rlist_add_tail_entry(&space->coll_id_holders, h, in_space);
+	coll_id_pin(coll_id, h, holder_type);
+}
+
+void
+space_pin_collations(struct space *space)
+{
+	struct tuple_format *format = space->format;
+	for (uint32_t i = 0; i < tuple_format_field_count(format); i++) {
+		struct tuple_field *field = tuple_format_field(format, i);
+		if (field->coll_id == COLL_NONE)
+			continue;
+		space_pin_collations_helper(space, field->coll_id,
+					    COLL_ID_HOLDER_SPACE_FORMAT);
+	}
+
+	for (uint32_t i = 0; i < space->index_count; i++) {
+		struct key_def *key_def = space->index[i]->def->key_def;
+		for (uint32_t i = 0; i < key_def->part_count; i++) {
+			struct key_part *part = &key_def->parts[i];
+			if (part->coll_id == COLL_NONE)
+				continue;
```

Comment:
Nit: The `COLL_NONE` check could be moved to `space_pin_collations_helper`.

Reply:
Done.